distributed/comm/ucx.py

```diff
@@ -94,67 +99,92 @@ def peer_address(self) -> str:
-
-            for frame in frames:
-                await self.ep.send_obj(frame)
-            return sum(map(nbytes, frames))
+        )
+        await self.ep.send(np.array([nbytes(f) for f in frames], dtype=np.uint64))
+        # Send frames
+        for frame in frames:
+            if nbytes(frame) > 0:
+                if hasattr(frame, "__array_interface__") or hasattr(
+                    frame, "__cuda_array_interface__"
+                ):
+                    await self.ep.send(frame)
+                else:
+                    await self.ep.send(np.frombuffer(frame, dtype=np.uint8))
+        return sum(map(nbytes, frames))
 
     async def read(self, deserializers=("cuda", "dask", "pickle", "error")):
-        with log_errors():
-            if deserializers is None:
-                deserializers = ("cuda", "dask", "pickle", "error")
-            resp = await self.ep.recv_future()
-            obj = ucp.get_obj_from_msg(resp)
-            (nframes,) = struct.unpack(
-                "Q", obj[:8]
-            )  # first eight bytes for number of frames
-
-            gpu_frame_msg = obj[
-                8 : 8 + nframes
-            ]  # next nframes bytes for if they're GPU frames
-            is_gpus = struct.unpack("{}?".format(nframes), gpu_frame_msg)
-
-            sized_frame_msg = obj[8 + nframes :]  # then the rest for frame sizes
-            sizes = struct.unpack("{}Q".format(nframes), sized_frame_msg)
+        if self.closed():
+            raise CommClosedError
 
-            frames = []
+        if deserializers is None:
+            deserializers = ("cuda", "dask", "pickle", "error")
 
-            for i, (is_gpu, size) in enumerate(zip(is_gpus, sizes)):
+        try:
+            # Recv meta data
+            nframes = np.empty(1, dtype=np.uint64)
+            await self.ep.recv(nframes)
+            is_cudas = np.empty(nframes[0], dtype=np.bool)
+            await self.ep.recv(is_cudas)
+            sizes = np.empty(nframes[0], dtype=np.uint64)
+            await self.ep.recv(sizes)
+        except (ucp.exceptions.UCXCanceled, ucp.exceptions.UCXCloseError):
+            if self._ep is not None and not self._ep.closed():
+                await self._ep.shutdown()
+                self._ep.close()
+            self._ep = None
+            raise CommClosedError()
+        else:
+            # Recv frames
+            frames = []
+            for is_cuda, size in zip(is_cudas.tolist(), sizes.tolist()):
                 if size > 0:
-                    resp = await self.ep.recv_obj(size, cuda=is_gpu)
+                    if is_cuda:
+                        frame = numba.cuda.device_array((size,), dtype=np.uint8)
+                    else:
+                        frame = np.empty(size, dtype=np.uint8)
+                    await self.ep.recv(frame)
+                    if is_cuda:
+                        frames.append(frame)
+                    else:
+                        frames.append(frame.tobytes())
```

Comment:
I wonder if we could pass the memoryview `frame.data`, here instead.

Reply:
Fixed